// Copyright 2016 The Cartographer Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";

package cartographer.mapping.constraints.proto;

import "cartographer/mapping/proto/scan_matching/ceres_scan_matcher_options_2d.proto";
import "cartographer/mapping/proto/scan_matching/global_icp_scan_matcher_options_2d.proto";

message ConstraintBuilderOptions {
  // Threshold for the scan match score below which a match is not considered.
  // Low scores indicate that the scan and map do not look similar.
  double min_local_search_score = 1;

  // Threshold below which global localizations are not trusted.
  double min_global_search_score = 2;

  // Weight used in the optimization problem for the translational component of loop closure constraints.
  double constraint_translation_weight = 3;

  // Weight used in the optimization problem for the rotational component of loop closure constraints.
  double constraint_rotation_weight = 4;

  // Options for the internally used scan matcher
  mapping.scan_matching.proto.CeresScanMatcherOptions2D ceres_scan_matcher_options = 6;
  mapping.scan_matching.proto.GlobalICPScanMatcherOptions2D global_icp_scan_matcher_options_2d = 7;

  // Minimum score to consider an ICP scan match successful
  double min_icp_score = 8;
  double min_icp_points_inlier_fraction = 9;
  double min_icp_features_inlier_fraction = 10;

  // Minimum scan agreement required for a successful scan match
  double min_hit_fraction = 11;
  double min_ray_trace_fraction = 12;
  double min_icp_features_match_fraction = 13;
}
